src/osdc/Journaler.h

```diff
@@ -393,6 +393,7 @@ class Journaler {
 
   uint32_t write_iohint; // the fadvise flags for write op, see
 			 // CEPH_OSD_OP_FADIVSE_*
+  std::mutex replaylock;
```

Comment:
Is the `replaylock` really required? Seems to me that its not protecting anything that the existing lock (`lock`) is protecting.

Reply:
@vshankar 
1, We know that if " readable_waiter.wait()" is called, the complete() must be called whenever it is before or after "readable_waiter.wait()" is called. Or it will hang here.
2, If we donot have replaylock, how the problem happened?
step(1):md_log_replay switch before "wait_for_readable(&readable_waiter)"(L1621).
step(2):on finisher thread, in _assimilate_prefetch(): in partial fragment case, "readable = _is_readable()" may set Journaler::readable to false; and the f->complete() was not called!!( here Journaler::on_readable is false.)
step(3):switch to md_log_replay. In " wait_for_readable(&readable_waiter)", "on_readable = wrap_finisher(onreadable)" will be called because readable is false.
step(4):hanged on "readable_waiter.wait()".
3,If we use replaylock:
(1)When we call is_readable() on line 1617, because it hold replaylock, _finish_read will can not be executed.
(2)when wait_for_readable(&readable_waiter) is executed and the replaylock is unlocked (line 1621 and 1622), _finish_read can get replaylock and can be executed.
(3)In _assimilate_prefetch, because the Journaler::on_readable is not NULL, the "f->complete()" will always can be called.
Then is will not hang here.
4,https://github.com/ceph/ceph/pull/48454
(1)This issue change the function wait_for_readable, which is used in several functions, for example: PurgeQueue::_consume(in which it maybe called in different condition).
(2)Logically, if use "finisher->queue(onreadable, 0)", "readable_waiter.wait()" will lost the return value.
We had reviewed this change innerally and we are in a team,haha.

Reply:
I understand the problem with the current code. What I meant was if we could use the existing `lock`. Something like

```
int Journaler::check_isreadable()
{
  int r=0;
  std::unique_lock l(lock);
  while (!is_readable() &&
      get_read_pos() < get_write_pos() &&
      !get_error()) {
    C_SaferCond readable_waiter;
    wait_for_readable(&readable_waiter);
    l.unlock();
    r = readable_waiter.wait();
    l.lock();
  }
  return r;
}
```

Would that solve the infinite wait problem?

Reply:
1, In function is_readable(), it has used the Journaler::lock.
2，is_readable() had been called by multiple functions.
3, If we take the lock out from is_readable(), we should add lock outside all the places where is_readable() is called.(Because "bool r = readable;" and "_prefetch();" in is_readable() should be locked on the outside )

Reply:
Hmm.. There is `Journaler::_is_readable()` which is pretty different than `Journaler::is_readable()`. How about having a version of is_readable() that can be called while holding `Journaler::lock`. Would that suffice?

Reply:
1, A new Journaler::_wait_for_readable not holding Journaler::lock is needed too.
2, By the way, what is the name for new is_readable not holding Journaler::lock, do you have any suggestions?
Maybe we can use this method, it donot need to import replaylock.

Reply:
> 1, A new Journaler::_wait_for_readable not holding Journaler::lock is needed too.

Of course.

> 2, By the way, what is the name for new is_readable not holding Journaler::lock, do you have any suggestions? Maybe we can use this method, it donot need to import replaylock.

`_is_readable` would have been ideal. Maybe call the current `_is_readable` to `_have_next_entry` and implement `_is_readable` that does not acquire `Journaler::lock`?